### ISSUE TITLE
Add `--interactive --all` CLI flags in `meltano config set`

### ISSUE THREAD

**Issue by MeltyBot** [meltano/meltano#3153]
Migrated from GitLab: https://gitlab.com/meltano/meltano/-/issues/3227

Originally created by @aaronsteers on 2022-02-09 21:01:33

---

Based on recent discussions around the difficulties in knowing _which_ settings to configure when adding a new plugin, we should add an interactive mode for setting all setting for a plugin, or any specific settings value.

Something like...

```console
> meltano config dbt-snowflake set --interactive --all

You are now configuring the "dbt-snowflake" transformer plugin.
For help, please refer to plugin documentation: https://...

We will now step through each setting. Leave any entry blank to skip, 
or enter a new value. Password values will automatically be hidden 
from display for any settings of type 'password'.

Press Ctrl+C to cancel at any time.

Setting #1: project_dir
Description: The root dbt path.
Type: string
Default value: $MELTANO_PROJECT_ROOT/transform
Current value: (none)
New Value: _ 

(skipped)

....

Setting #7: snowflake_user
Description: The username to use in Snowflake authentication.
Type: string
Default value: (none)
Current value: (none)
New Value: snowflake_prod_user

Value saved successfully to meltano.yml!

Setting #8: snowflake_password
Description: The password to use in Snowflake authentication.
Type: password
Default value: $MELTANO_PROJECT_ROOT/transform
Current value: (none)
New Value: ********** 

Value saved successfully to Meltano System DB!

```

## Spec Details

- Interactive behavior is triggered bia the `--interactive` flag in `meltano config ... set`.
- Interactive behavior can be triggered for a single setting or with `--all` to set all settings associated with the given plugin.
- The `--all` option _requires_ the `--interactive` flag. If `--all` is specified without `--interactive`, Meltano should fail.
- After each setting is set, Meltano should respond with `Value saved successfully to _<store>_!`
  - The store could be `.env`, the SystemDB, or to Meltano.yml.
  - Confirming and saving after each setting is set gives real-time feedback to the user about changes they have applied, and allows the remaining operation to be canceled without loss of any setting values already saved.
- Individual settings should be skippable.
- Password fields should either print asterisks or simply not echo anything while users are typing.
- The new setting value should not be provided inline at the command line if `--interactive` is set. So, passing the setting value and also passing `--interactive` should cause Meltano to fail.

## Ability to "skip" settings

Skipping a setting may not be intuitive if done inline with the setting value input.

```
Setting #1: project_dir
Description: The root dbt path.
Type: string
Default value: $MELTANO_PROJECT_ROOT/transform
Current value: (none)
Would you like to set a new value? [Y/n]: n

Skipped.

Setting #2: snowflake_user
Description: The username to use in Snowflake authentication.
Type: string
Default value: (none)
Current value: (none)
Would you like to set a new value? [Y/n]: Y
New Value? []: snowflake_prod_user

Value saved successfully to meltano.yml!

...
```

Without a separate prompt to skip or set the value, the user would generally have to type something like an empty string or `s` for skip, or similar. However, those are potentially valid answers to the new value prompt.

## Phases

- The initial merge should at minimum provide `--all --interactive` support.
- Optionally, setting only one value (`--interactive` without `--all`) can be included in initial scope or added in a subsequent iteration.
  - When setting only a single value, presumably we do not need to allow the "skip" option described above, since the user can simply abort (`ctrl+c`) to cancel if they do not want the value set.
- Not required in this first iteration, but a future iteration should require users to first define a setting's type if the type is not yet set. Would be a path for resolving [#2768](https://gitlab.com/meltano/meltano/-/issues/2768). For example, if the user runs `meltano config tap-snowflake set --interactive snowflake_password`, that should require the user to _first_ define the setting `kind` (kind is `password` in this case), which then drives correct parsing and storage of the provided value.

**Comment by MeltyBot** [comment]
[View 29 previous comments from the original issue on GitLab](https://gitlab.com/meltano/meltano/-/issues/3227)

**Comment by kgpayne** [comment]
@aaronsteers @tayloramurphy @pnadolny13 (and @pandemicsyn who has formatting experience from logging)

As a consequence of pretty formatting in Setting description fields for the hub, we now have unformatted artefacts appearing in `interactive` mode. To address these, we can:

1. Just live with it. They aren't ideal, but not the end of the world.
2. Use a library like [rich](https://github.com/Textualize/rich) which supports printing [Markdown to the terminal](https://rich.readthedocs.io/en/stable/markdown.html).
3. Find a library to strip html/markdown and other formatting from description strings.
4. Add a new `short_description` field to our Setting schema for descriptions without formatting.

Examples of current artefacts:

<img width="715" alt="Screenshot 2022-06-13 at 14 52 32" src="https://user-images.githubusercontent.com/5585874/173375663-94b1bff4-2fcb-4ff6-897f-1e609dfd1962.png">

<img width="710" alt="Screenshot 2022-06-13 at 14 53 43" src="https://user-images.githubusercontent.com/5585874/173375691-df45579b-954e-4e60-9188-f94e32437fa9.png">

**Comment by kgpayne** [comment]
I also opened #6172 to track another bug that came up in my testing 🙂

**Comment by tayloramurphy** [comment]
@kgpayne for this issue, I'm ok with leaving it as is since we have the other issues that track updating the printing experience on the CLI. 

For Option 4, I'd rather just have an ellipsis after the first sentence or first N characters. Managing a completely separate key is unlikely to be scalable.

**Comment by kgpayne** [comment]
@aaronsteers (FYI @tayloramurphy), notes from our 1:1 yesterday:

In testing `meltano config tap-gitlab set --interactive` I came across an example setting description (`tap-gitlab` setting `private_token`) that is long and includes Markdown formatting. This does't look great as plain text (spot the embedded video 😅):

![Image](https://user-images.githubusercontent.com/5585874/185431751-48ec0924-bdd3-41a8-9457-76d59cea7075.png)

In talking with AJ, we came up with a few options. Naturally there may be more, but we talked briefly about:

1. Continuing anyway, knowing the experience won't be perfect, and fixing this in the hub yaml files.
2. Add [markdown rendering](https://rich.readthedocs.io/en/stable/markdown.html) with `rich`, after #2971 merges.
3. Strip the markdown elements before printing the description field.
4. Introduce a new `short_description` or similar for plaintext description key that does not allow markdown. AJ also mentioned a new `usage` key (already implemented) that might be more appropriate for rich text content. We also discussed printing a URL to the homepage/repo with a prompt along the lines of "for detailed help, see the plugins' documentation" to compliment a shortened description.

At the end of our call, we resolved to evaluate 2. by rendering the [example description](https://github.com/meltano/hub/blob/main/_data/meltano/extractors/tap-gitlab/meltanolabs.yml#L32) with rich. It turned out as follows:

![Image](https://user-images.githubusercontent.com/5585874/185435209-7da86b8a-242f-436b-86ad-37bc73eae545.png)

It would be helpful to know if this is good enough to unblock the `rich` addition, or if we'd prefer to have a longer discussion around rich docs in plugin definitions 🙂

**Comment by tayloramurphy** [comment]
@kgpayne option 2 is a good iteration and I like AJ's comments in https://github.com/meltano/meltano/issues/2971#issuecomment-1212326558 on having it behind a feature flag and configurable. 

I can open a new issue on the possibility of a `short_description` key for settings. Note that the `usage` key is for the entire plugin and not a specific setting.

**Comment by aaronsteers** [comment]
@kgpayne - Yes, I agree option 2 is great for this phase.

We want to be careful that the addition of `rich` doesn't automatically change other behaviors, but once the feature flag is in place for #2971, that should be safe to add.

Also - thanks for the rendering example. I think it looks nice and is worthwhile having markdown-rendering-to-terminal as a tool in our toolbox overall.
